File: Examples/NativeExample/NativeExample/ViewController.m
/*
 *  ViewController.m
 *
 * This file is a part of the Yandex Advertising Network.
 *
 * Version for iOS © 2016 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */

#import "ViewController.h"
#import "NativeAppInstallAdView.h"
#import "NativeContentAdView.h"
#import <YandexMobileAds/YandexMobileNativeAds.h>

@interface ViewController () <YMANativeAdLoaderDelegate, YMANativeAdDelegate>

@property (nonatomic, strong) NativeContentAdView *contentAdView;
@property (nonatomic, strong) NativeAppInstallAdView *appInstallAdView;
@property (nonatomic, strong) YMANativeAdLoader *adLoader;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.contentAdView = [[NativeContentAdView alloc] initWithFrame:CGRectZero];
    self.contentAdView.translatesAutoresizingMaskIntoConstraints = NO;
    self.contentAdView.backgroundColor = [UIColor yellowColor];

    self.appInstallAdView = [[NativeAppInstallAdView alloc] initWithFrame:CGRectZero];
    self.appInstallAdView.translatesAutoresizingMaskIntoConstraints = NO;
    self.appInstallAdView.backgroundColor = [UIColor colorWithWhite:0.9f alpha:1.f];

    // Replace demo R-M-DEMO-native-c with actual Block ID
    // Following demo Block IDs may be used for testing:
    // R-M-DEMO-native-c
    // R-M-DEMO-native-i
    YMANativeAdLoaderConfiguration *configuration =
        [[YMANativeAdLoaderConfiguration alloc] initWithBlockID:@"R-M-DEMO-native-c"
                                        loadImagesAutomatically:YES];
    self.adLoader = [[YMANativeAdLoader alloc] initWithConfiguration:configuration];
    self.adLoader.delegate = self;
    [self.adLoader loadAdWithRequest:nil];
}

- (void)removeCurrentAdView
{
    [self.appInstallAdView removeFromSuperview];
    [self.contentAdView removeFromSuperview];
}

- (void)addConstraintsToAdView:(UIView *)adView
{
    NSDictionary *views = NSDictionaryOfVariableBindings(adView);
    NSArray *vertical = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[adView]|"
                                                                options:0
                                                                metrics:nil
                                                                  views:views];
    NSArray *horizontal = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[adView]|"
                                                                  options:0
                                                                  metrics:nil
                                                                    views:views];
    NSArray *constraints = [vertical arrayByAddingObjectsFromArray:horizontal];
    [self.view addConstraints:constraints];
}

#pragma mark - YMANativeAdLoaderDelegate

- (void)nativeAdLoader:(null_unspecified YMANativeAdLoader *)loader
   didLoadAppInstallAd:(id<YMANativeAppInstallAd> __nonnull)ad
{
    [self removeCurrentAdView];

    NSError * __autoreleasing error = nil;
    [ad bindAppInstallAdToView:self.appInstallAdView delegate:self error:&error];
    NSLog(@"Binding finished with error: %@", error);

    [self.appInstallAdView prepareForDisplay];

    [self.view addSubview:self.appInstallAdView];
    [self addConstraintsToAdView:self.appInstallAdView];
}

- (void)nativeAdLoader:(null_unspecified YMANativeAdLoader *)loader
      didLoadContentAd:(id<YMANativeContentAd> __nonnull)ad
{
    [self removeCurrentAdView];

    NSError * __autoreleasing error = nil;
    [ad bindContentAdToView:self.contentAdView delegate:self error:&error];
    NSLog(@"Binding finished with error: %@", error);

    [self.contentAdView prepareForDisplay];

    [self.view addSubview:self.contentAdView];
    [self addConstraintsToAdView:self.contentAdView];
}

- (void)nativeAdLoader:(null_unspecified YMANativeAdLoader *)loader didFailLoadingWithError:(NSError * __nonnull)error
{
    NSLog(@"Native ad loading error: %@", error);
}

#pragma  mark - YMANativeAdDelegate

// Uncomment to open web links in in-app browser

//- (UIViewController *)viewControllerForPresentingModalView
//{
//    return self;
//}

- (void)nativeAdWillLeaveApplication:(null_unspecified id)ad
{
    NSLog(@"Will leave application");
}

- (void)nativeAd:(null_unspecified id)ad willPresentScreen:(UIViewController *)viewController
{
    NSLog(@"Will present screen");
}

- (void)nativeAd:(null_unspecified id)ad didDismissScreen:(UIViewController *)viewController
{
    NSLog(@"Did dismiss screen");
}

@end

File: Examples/Swift/NativeExample/NativeExample/ViewController.swift
/*
 * Version for iOS © 2015–2019 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://yandex.com/legal/mobileads_sdk_agreement/
 */

import YandexMobileAds

class ViewController: UIViewController {
    private var contentAdView: NativeContentAdView?
    private var appInstallView: NativeAppInstallAdView?
    private var imageAdView: NativeImageAdView?
    
    var adLoader: YMANativeAdLoader!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadAdViews()
        hideAllViews()
        
        // Replace demo R-M-DEMO-native-c with actual Block ID
        // Following demo Block IDs may be used for testing:
        // R-M-DEMO-native-c
        // R-M-DEMO-native-i
        // R-M-DEMO-native-video
        let configuration = YMANativeAdLoaderConfiguration(blockID: "R-M-DEMO-native-c",
                                                           loadImagesAutomatically: true)
        
        adLoader = YMANativeAdLoader(configuration: configuration)
        adLoader.delegate = self
        adLoader.loadAd(with: nil)
    }
    
    private func loadAdViews() {
        contentAdView = Bundle.main.loadNibNamed("NativeContentAdView",
                                                 owner: nil,
                                                 options: nil)?.first as? NativeContentAdView
        if let myContentAdView = contentAdView {
            addAdView(myContentAdView)
        }
        
        appInstallView = Bundle.main.loadNibNamed("NativeAppInstallAdView",
                                                  owner: nil,
                                                  options: nil)?.first as? NativeAppInstallAdView
        if let myAppInstallView = appInstallView {
            addAdView(myAppInstallView)
        }
        
        imageAdView = Bundle.main.loadNibNamed("NativeImageAdView",
                                               owner: nil,
                                               options: nil)?.first as? NativeImageAdView
        if let myImageAdView = imageAdView {
            addAdView(myImageAdView)
        }
    }
    
    private func addAdView(_ adView: UIView) {
        adView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(adView)
        let views = ["adView": adView]
        let horizontal = NSLayoutConstraint.constraints(withVisualFormat: "H:|-[adView]-|",
                                                        options: [],
                                                        metrics: nil,
                                                        views: views)
        let vertical = NSLayoutConstraint.constraints(withVisualFormat: "V:[adView]-|",
                                                      options: [],
                                                      metrics: nil,
                                                      views: views)
        view.addConstraints(horizontal + vertical)
    }
    
    private func hideAllViews() {
        imageAdView?.isHidden = true
        appInstallView?.isHidden = true
        contentAdView?.isHidden = true
    }
    
    private func showAdView(_ adView: UIView) {
        adView.isHidden = false
    }
}

// MARK: - YMANativeAdLoaderDelegate

extension ViewController: YMANativeAdLoaderDelegate {
    func nativeAdLoader(_ loader: YMANativeAdLoader!, didLoad ad: YMANativeImageAd) {
        guard let myImageAdView = imageAdView else { return }
        try? ad.bindImageAd(to: myImageAdView, delegate: self)
        showAdView(myImageAdView)
        myImageAdView.configureAssetViews()
    }
    
    func nativeAdLoader(_ loader: YMANativeAdLoader!, didLoad ad: YMANativeContentAd) {
        guard let myContentAdView = contentAdView else { return }
        try? ad.bindContentAd(to: myContentAdView, delegate: self)
        showAdView(myContentAdView)
        myContentAdView.configureAssetViews()
    }
    
    func nativeAdLoader(_ loader: YMANativeAdLoader!, didLoad ad: YMANativeAppInstallAd) {
        guard let myAppInstallView = appInstallView else { return }
        try? ad.bindAppInstallAd(to: myAppInstallView, delegate: self)
        showAdView(myAppInstallView)
        myAppInstallView.configureAssetViews()
    }
    
    func nativeAdLoader(_ loader: YMANativeAdLoader!, didFailLoadingWithError error: Error) {
        print("Native ad loading error: \(error)")
    }
}

// MARK: - YMANativeAdDelegate

extension ViewController: YMANativeAdDelegate {
    func closeNativeAd(_ ad: Any!) {
        print("Close native ad")
        hideAllViews()
    }
}
